File: BeakjoonOnlineJudge/2740/2740.swift
/*
https://www.acmicpc.net/problem/2740
BOJ 2740번 행렬곱

(1) 설명
스위프트로 C++과 동일한 방식으로 구현하였다.
하지만 채점 방식의 입력과 문제가 발생하여 런타임에러가 난다. 
알고리즘을 동일하게 구현했다는 것에 의의를 둔다.

스위프트로 백준의 문제를 풀어본 적이 처음이기 때문에
채점을 통과하지 못한 것이 아쉽지만, 스위프트를 조금 더 이해할 수 있었다.
*/
import Foundation

func input(_ A : inout Int, _ B : inout Int)
{
    let inputLine = readLine()
    guard let _ = inputLine else {
        return
    }
    let array = inputLine?.components(separatedBy: " ")
    if let num = Int(array?[0] ?? "0") {
        A = num
    }
    if let num = Int(array?[1] ?? "0") {
        B = num
    }
}

func setMatrix(_ matrix : inout [[Int]], row : Int, column : Int)
{
    matrix = Array(repeating: Array(repeating: 0, count: column), count: row)
    for i in 0..<row
    {
        let inputLine = readLine()
        guard let _ = inputLine else {
            return
        }
        let array = inputLine?.components(separatedBy: " ")
        for j in 0..<column
        {
            if let num = array?[j] {
                matrix[i][j] = Int(num) ?? 0
            }
        }
    }
}

func multipleMatrix(_ A : [[Int]], _ B : [[Int]]) -> [[Int]]
{
    var ans : [[Int]] = Array(repeating: Array(repeating: 0, count: N), count: K)
    for i in 0..<N
    {
        for j in 0..<K
        {
            for m in 0..<M
            {
                ans[i][j] += (A[i][m] * B[m][j])
            }
        }
    }
    return ans
}

func printMatrix(_ matrix : [[Int]])
{
    var forPrint : String = ""
    for i in 0..<matrix.count
    {
        for j in 0..<matrix[0].count
        {
            forPrint += String(matrix[i][j]) + " "
        }
        print(forPrint)
        forPrint = ""
    }
}

var N : Int = 0
var M : Int = 0
var K : Int = 0
var A : [[Int]] = Array(repeating: Array(repeating: 0, count: 0), count: 0)
var B : [[Int]] = Array(repeating: Array(repeating: 0, count: 0), count: 0)
var Ans : [[Int]] = Array(repeating: Array(repeating: 0, count: 0), count: 0)
input(&N, &M)
setMatrix(&A, row: N, column: M)
input(&M, &K)
setMatrix(&B, row: M, column: K)
Ans = multipleMatrix(A, B)
printMatrix(Ans)

File: BeakjoonOnlineJudge/2740/2740.cpp
/*
https://www.acmicpc.net/problem/2740
BOJ 2740번 행렬곱
(1) 접근 
행렬을 곱하는 문제이다. 
행렬을 곱하는 방법을 그대로 구현하면 된다.
(2) 해결
N * M 행렬과 M * K 행렬을 서로 곱하면 N * K의 행렬이 된다.
그러므로 행렬의 인덱스를 반복문을 이용하여 행렬 곱셈식 방법대로 각 행과 열의 곱셈결과를 더해주면
행렬의 곱셈을 완료할 수 있다.
*/
#include <iostream>

using namespace std;

int N, M, K;
int A[100][100];
int B[100][100];
int ANS[100][100];
int main()
{
    cin >> N >> M;
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            cin >> A[i][j];
        }
    }
    cin >> M >> K;
    for(int i = 0; i < M; i++)
    {
        for(int j = 0; j < K; j++)
        {
            cin >> B[i][j];
        }
    }
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < K; j++)
        {
            for(int k = 0; k < M; k++)
            {
                ANS[i][j] += A[i][k] * B[k][j];  
            }
        }
    }

    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < K; j++)
        {
            cout << ANS[i][j] << " ";
        }
        cout << "\n";
    }
    
}